docs/debug.md

```diff
@@ -104,8 +105,122 @@ Breakpoint 1, tee_entry_std (smc_args=0x7df6ff98 <stack_thread+8216>)
+                 CONTENTS, ALLOC, LOAD, DATA
+   1   .text     0000c614  0000000000000020  0000000000000020  00010020  2**3
+                 CONTENTS, ALLOC, LOAD, READONLY, CODE
+   2   .rodata   00002382  000000000000c638  000000000000c638  0001c638  2**3
+```
+
+We are interested in the LMA (Load Memory Address) of `.text` segment. It’s the
+address where the elf will be loaded to. 
+
+Remember, TA is loaded dynamically at runtime by kernel driver `tee_supplicant`
+when CA is requesting access to TA. That’s why LMA by itself cannot be used to
+determine the actual address of the TA loaded into OP-TEE OS. But we will need
+this value later on.
+
+Now, you can run gdb and connect to QEMU session.
+
+```
+(gdb) optee
+SIGTRAP is used by the debugger.
+Are you sure you want to change it? (y or n) [answered Y; input not from
+terminal]
+0x00000000 in ?? ()
+```
+
+We have to set a breakpoint somewhere inside OP-TEE OS. For example:
+```
+(gdb) b tee_ta_invoke_command
+Breakpoint 1 at 0xe111880: file core/kernel/tee_ta_manager.c, line 598.
+```
+
+then continue the execution of vm:
+```
+(gdb) c
+Continuing.
+```
+
+Now you can run Hello World CA (Client Application) from Normal World FVP terminal:
+```bash
+root@FVP:/ hello_world
+```
+
+`gdb` will hit the breakpoint at `tee_ta_invoke_command`, which will look something
+like this:
+```
+Breakpoint 1, tee_ta_invoke_command (err=0xe173890 <stack_thread+8016>, err@entry=0xe1738d0 <stack_thread+8080>,
+    sess=sess@entry=0xe163e10, clnt_id=clnt_id@entry=0x0, cancel_req_to=cancel_req_to@entry=4294967295, cmd=0,
+    param=0xe1738a8 <stack_thread+8040>, param@entry=0xe1738e8 <stack_thread+8104>) at core/kernel/tee_ta_manager.c:598
+598     {
+(gdb)
+```
+
+In the Secure World terminal window you should be able to see something like this:
+```
+DEBUG: [0x0] TEE-CORE:ta_load:316: ELF load address 0x40001000
+DEBUG: USER-TA:TA_CreateEntryPoint:43: has been called
+FLOW:  USER-TA: tee_user_mem_alloc:343: Allocate: link:[0x400190b0], buf:[0x400190d0:32]
+DEBUG: USER-TA:TA_OpenSessionEntryPoint:81: Hello World!
+```
+
+Here, we are interested in ELF load address value, in this case it is `0x40001000`.
+This is the actual address of Hello World TA loaded by the kernel driver.
+
+To be able to debug Hello World TA you have to add symbols of the application
+into `gdb` using `add-symbol-file [file] [address]` command, were `[address]` is
+ELF load address + LMA of `.text` segment. In this case, it's `0x40001000 + 0x20`
+
+Let’s add those symbols:
+```
+(gdb) add-symbol-file ./8aaaf200-2450-11e4-abe2-0002a5d5c51b.elf 0x40001020
+```
+
+You should see something like:
+```
+add symbol table from file "./8aaaf200-2450-11e4-abe2-0002a5d5c51b.elf" at .text_addr = 0x40001020
+```
+
+Now, we can setup a breakpoint inside Hello World TA, for example:
+```
+(gdb) b TA_InvokeCommandEntryPoint
+Breakpoint 2 at 0x400028e0: file hello_world_ta.c, line 135.
+```
+
+Then continue the execution of vm:
```

Comment:
Here one can remove the optee core breakpoint (`thread_enter_user_mode()` or `tee_ta_invoke_command()`)

Reply:
Agreed!